### PR TITLE
refactor(core): don't depend on get_identity_hash for modules

### PR DESCRIPTION
**Pull request by bartlomieju**
Closes https://github.com/denoland/deno/issues/7969

### CODE REVIEW

File: core/modules.rs [hunk 1]
```diff
@@ -468,11 +465,23 @@ impl Modules {
...
+    // FIXME(@bartlomieju): iterating over all registered modules, this
+    // should use hash map lookup
+    for (id, handle) in self.handles.iter() {
+      if handle == global {
+        return self.info.get(id);
+      }
     }
-    self.info.get(&id)
+
+    None
```
Comment (bartlomieju) [inline core/modules.rs:484]: This is not optimal... but to use hash lookup, we'd need to have both `HashMap<ModuleId, v8::Global<v8::Module>>` and `HashMap<v8::Global<v8::Module>, ModuleId>`. Not sure if it's worth it, OTOH, cloning a global should be very cheap, so maybe it makes sense after all. @bnoordhuis @piscisaureus opinions?

Comment (bartlomieju) [reply core/modules.rs]: Actually I just added another hash map for this mapping.

Comment (bartlomieju) [reply core/modules.rs]: Actually not a feasible solution, getting a panic:
```
#
# Fatal error in ../../../../v8/src/objects/objects-inl.h, line 1043
# Debug check failed: object.IsJSReceiver().
#
#
#
#FailureMessage Object: 0x70000dd8e840
==== C stack trace ===============================

         0x000000010d0b7e4a v8__internal__Object__GetHash + 122
    7   deno_core-631d91027172ddfc          0x000000010d087711 rusty_v8::data::Data::get_hash::h7452fb114852f76d + 17
    8   deno_core-631d91027172ddfc          0x000000010cf1cd96 _ZN59_$LT$rusty_v8..data..Module$u20$as$u20$core..hash..Hash$GT$4hash17h53f2481da618ebd1E + 38
    9   deno_core-631d91027172ddfc          0x000000010cefc679 _ZN70_$LT$rusty_v8..handle..Global$LT$T$GT$$u20$as$u20$core..hash..Hash$GT$4hash17h7dd474cc65a1fdc5E + 121
    10  deno_core-631d91027172ddfc          0x000000010cf3a7b8 hashbrown::map::make_hash::h31a6e439bd44ea82 + 72
    11  deno_core-631d91027172ddfc          0x000000010cf39166 _ZN9hashbrown3map24HashMap$LT$K$C$V$C$S$GT$6insert17hb6261f499a1fbb28E + 70
    12  deno_core-631d91027172ddfc          0x000000010cee4e9c _ZN3std11collections4hash3map24HashMap$LT$K$C$V$C$S$GT$6insert17he9074b15704ccf0cE + 28
    13  deno_core-631d91027172ddfc          0x000000010cf79996 deno_core::modules::Modules::register::h900f8583d8144714 + 342
    14  deno_core-631d91027172ddfc          0x000000010cf6a88b deno_core::runtime::JsRuntime::mod_new::h1dfec95196c0c777 + 3147
    15  deno_core-631d91027172ddfc          0x000000010cf6fd74 deno_core::runtime::JsRuntime::register_during_load::h50f5a5dc2827dc7d + 1716
    16  deno_core-631d91027172ddfc          0x000000010cf46362 _ZN9deno_core7runtime9JsRuntime11load_module28_$u7b$$u7b$closure$u7d$$u7d$17h6bfccbd8b2e191c7E + 2610
    17  deno_core-631d91027172ddfc          0x000000010cefb5ab _ZN97_$LT$core..future..from_generator..GenFuture$LT$T$GT$$u20$as$u20$core..future..future..Future$GT$4poll17h965bd7fe34e2d42dE + 107
    18  deno_core-631d91027172ddfc          0x000000010cf5b5be _ZN9deno_core7modules5tests18test_redirect_load28_$u7b$$u7b$closure$u7d$$u7d$17hfb42e9a578d03bc8E + 462
    19  deno_core-631d91027172ddfc          0x000000010cefbab0 _ZN97_$LT$core..future..from_generator..GenFuture$LT$T$GT$$u20$as$u20$core..future..future..Future$GT$4poll17hff946ba0814b26c6E + 80
    20  deno_core-631d91027172ddfc          0x000000010cfdf74e _ZN72_$LT$core..pin..Pin$LT$P$GT$$u20$as$u20$core..future..future..Future$GT$4poll17h63d63e0d4c738324E + 94
    21  deno_core-631d91027172ddfc          0x000000010cf3df30 _ZN16futures_executor10local_pool8block_on28_$u7b$$u7b$closure$u7d$$u7d$17h7e0eedfb4a24a9e6E + 64
    22  deno_core-631d91027172ddfc          0x000000010cf3c3a2 _ZN16futures_executor10local_pool12run_executor28_$u7b$$u7b$closure$u7d$$u7d$17h271a24406e401341E + 98
    23  deno_core-631d91027172ddfc          0x000000010ced9b51 _ZN3std6thread5local17LocalKey$LT$T$GT$8try_with17h80cf2a89f735aa0cE + 145
    24  deno_core-631d91027172ddfc          0x000000010ced9175 _ZN3std6thread5local17LocalKey$LT$T$GT$4with17hd354fae7002cadb2E + 21
    25  deno_core-631d91027172ddfc          0x000000010cf3bbe3 futures_executor::local_pool::run_executor::hd4f566e4a55cdc77 + 115
    26  deno_core-631d91027172ddfc          0x000000010cf3d6c5 futures_executor::local_pool::block_on::habb0b2d82d770ac2 + 69
    27  deno_core-631d91027172ddfc          0x000000010cf1b0a7 deno_core::modules::tests::test_redirect_load::hc0b6a5fe53e40473 + 503
    28  deno_core-631d91027172ddfc          0x000000010cf5b3e1 _ZN9deno_core7modules5tests18test_redirect_load28_$u7b$$u7b$closure$u7d$$u7d$17h30f6f3e5c90fa4deE + 17
    29  deno_core-631d91027172ddfc          0x000000010cf7df61 core::ops::function::FnOnce::call_once::h69ae7bded0e1b7ea + 17
    30  deno_core-631d91027172ddfc          0x000000010d00e27a test::__rust_begin_short_backtrace::h3525a66645c5bfc7 + 10
    31  deno_core-631d91027172ddfc          0x000000010d00caee _ZN4test8run_test14run_test_inner28_$u7b$$u7b$closure$u7d$$u7d$17h2b2c6792a791ad99E + 798
    32  deno_core-631d91027172ddfc          0x000000010cfe6b1b std::sys_common::backtrace::__rust_begin_short_backtrace::h25e344bd33c2e044 + 43
    33  deno_core-631d91027172ddfc          0x000000010cfebeb5 _ZN4core3ops8function6FnOnce40call_once$u7b$$u7b$vtable.shim$u7d$$u7d$17h002c6dc1d62df3bcE + 117
```

File: core/modules.rs [hunk 2]
```diff
@@ -468,11 +464,23 @@ impl Modules {
...
+  pub fn get_handle(&self, id: ModuleId) -> Option<v8::Global<v8::Module>> {
+    self.handles.get(&id).cloned()
+  }
+
+  pub fn get_info(
+    &self,
+    global: &v8::Global<v8::Module>,
+  ) -> Option<&ModuleInfo> {
+    // NOTE(@bartlomieju): not ideal, but I was unable to create
+    // another hashmap indexed by `handle`.
```
Comment (bnoordhuis) [inline core/modules.rs:476]: If (and only if) you remember: what was the error you got? `Global` implements `PartialEq` and `Hasher` so it should be possible to use them as keys.

**Review by bnoordhuis — APPROVED** [review]

File: core/modules.rs (hunk 2 above)
Comment (bartlomieju) [reply core/modules.rs:476]: I thought so too, here's the error: https://github.com/denoland/deno/pull/8354#discussion_r522157813

Comment (piscisaureus) [reply core/modules.rs:476]: That's because v8__Data__GetHash doesn't work right for Module handles.
cc @bnoordhuis

Comment (bartlomieju) [reply core/modules.rs:476]: @piscisaureus I think it was already fixed in https://github.com/denoland/rusty_v8/pull/522 just waiting on the release

File: core/modules.rs (hunk 1 above)
Comment (bartlomieju) [reply core/modules.rs]: Applied change once more

**Comment by bartlomieju** [comment]
Blocked by https://github.com/denoland/deno/pull/8446

**Pull request merged**